lib/Sema/CSDiagnostics.h

```diff
@@ -787,6 +787,33 @@ class MissingMemberFailure final : public FailureDiagnostic {
   static DeclName findCorrectEnumCaseName(Type Ty,
                                           TypoCorrectionResults &corrections,
                                           DeclName memberName);
+
+protected:
+  Type getBaseType() const { return BaseType; }
+  DeclName getName() const { return Name; }
+};
+
+/// Diagnose cases where a member only accessible on generic constraints
+/// requiring conformance to a protocol is used on a value of the
+/// existential protocol type e.g.
+///
+/// ```swift
+/// protocol P {
+///   var foo: Self { get }
+/// }
+///
+/// func bar<X : P>(p: X) {
+///   p.foo
+/// }
+/// ```
+class AllowProtocolTypeMemberFailure final : public MissingMemberFailure {
```

Comment:
I think the name has to get adjusted to something like `InvalidMemberRefOnExistential` or similar to point out what the problem is. Another small refactoring you can do is extract baseType/Name related fields into `InvalidMemberRefFailure` diagnostic and use it as a base class for both `MissingMember` and here.